cache/cache.go

```diff
@@ -0,0 +1,928 @@
+		return nil, errors.Wrapf(err, "failed to understand duration", metaAge)
+	}
+	warmupRps := strings.Split(*cacheWarmUp, "/")
+	warmupRate, err := strconv.Atoi(warmupRps[0])
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to understand warm up rate", *cacheWarmUp)
+	}
+	warmupSec, err := strconv.Atoi(warmupRps[1])
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to understand warm up seconds", *cacheWarmUp)
+	}
+	// configure cache backend
+	if *cacheDbPurge {
+		fs.Debugf(name, "Purging the DB")
+	}
+	f := &Fs{
+		Fs:                   wrappedFs,
+		name:                 name,
+		root:                 rpath,
+		fileAge:              infoDuration,
+		chunkSize:            int64(chunkSize),
+		chunkAge:             chunkDuration,
+		metaAge:              metaDuration,
+		readRetries:          *cacheReadRetries,
+		totalWorkers:         *cacheTotalWorkers,
+		originalTotalWorkers: *cacheTotalWorkers,
+		chunkMemory:          !*cacheChunkNoMemory,
+		originalChunkMemory:  !*cacheChunkNoMemory,
+		warmUp:               false,
+		warmUpRate:           warmupRate,
+		warmUpSec:            warmupSec,
+		cacheWrites:          *cacheStoreWrites,
+		lastChunkCleanup:     time.Now().Truncate(time.Hour * 24 * 30),
+		lastRootCleanup:      time.Now().Truncate(time.Hour * 24 * 30),
+		lastOpenedEntries:    make(map[string]time.Time),
+	}
+
+	f.rateLimiter = rate.NewLimiter(rate.Limit(float64(*cacheRps)), f.totalWorkers)
+
+	dbPath := *cacheDbPath
+	if path.Ext(dbPath) != "" {
+		dbPath = path.Dir(dbPath)
+	}
+	dbPath = path.Join(dbPath, name+".db")
+	fs.Infof(name, "Storage DB path: %v", dbPath)
+	f.cache = GetPersistent(dbPath, *cacheDbPurge)
+	if err != nil {
+		return nil, err
+	}
+
+	fs.Infof(name, "Chunk Memory: %v", f.chunkMemory)
+	fs.Infof(name, "Chunk Size: %v", fs.SizeSuffix(f.chunkSize))
+	fs.Infof(name, "Workers: %v", f.totalWorkers)
+	fs.Infof(name, "File Age: %v", f.fileAge.String())
+	fs.Infof(name, "Chunk Age: %v", f.chunkAge.String())
+	fs.Infof(name, "Cache Writes: %v", f.cacheWrites)
+
+	go f.CleanUpCache(false)
+
+	// TODO: Explore something here but now it's not something we want
+	// when writing from cache, source FS will send a notification and clear it out immediately
+	//setup dir notification
+	//doDirChangeNotify := wrappedFs.Features().DirChangeNotify
+	//if doDirChangeNotify != nil {
+	//	doDirChangeNotify(func(dir string) {
+	//		d := NewAbsDirectory(f, dir)
+	//		d.Flush()
+	//		fs.Infof(dir, "updated from notification")
+	//	}, time.Second * 10)
+	//}
+
+	f.features = (&fs.Features{
+		CaseInsensitive:         wrappedFs.Features().CaseInsensitive,
+		DuplicateFiles:          false, // storage doesn't permit this
+		ReadMimeType:            wrappedFs.Features().ReadMimeType,
+		WriteMimeType:           wrappedFs.Features().WriteMimeType,
+		CanHaveEmptyDirectories: wrappedFs.Features().CanHaveEmptyDirectories,
+		BucketBased:             wrappedFs.Features().BucketBased,
+		Purge:                   f.Purge,
```

Comment:
"Fill just fills in the feature methods" - yes that is what I meant by "You don't need to write the methods here - Fill will fill them for you." sorry if that wasn't clear.

> Mask will AND the feature flags and since by default they will be false, a lot of features that would be supported by the wrapped FS will end up as not supported really even though they might be.

Yes you set to true things you could support.